readers/mutation_reader.cc

```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
+        // will compare less than any other key, making sure we don't attempt to
+        // compare partition-keys belonging to different tokens.
+        if (!_prev_partition_key.key().is_empty()) {
+            _prev_partition_key._key = partition_key::make_empty();
+        }
     }
-    return false;
+    return true;
+}
+
+bool mutation_fragment_stream_validator::operator()(const dht::decorated_key& dk) {
+    return validate(dk.token(), &dk.key());
 }
 
 bool mutation_fragment_stream_validator::operator()(dht::token t) {
-    if (_prev_partition_key.token() <= t) {
-        _prev_partition_key._token = t;
-        return true;
-    }
-    return false;
+    return validate(t, nullptr);
 }
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind, position_in_partition_view pos) {
+bool mutation_fragment_stream_validator::validate(mutation_fragment_v2::kind kind, std::optional<position_in_partition_view> pos,
+    std::optional<tombstone> new_current_tombstone) {
+    // Check for unclosed range tombstone on partition end
     if (kind == mutation_fragment_v2::kind::partition_end && _current_tombstone) {
         return false;
     }
-    if (_prev_kind == mutation_fragment_v2::kind::partition_end) {
-        const bool valid = (kind == mutation_fragment_v2::kind::partition_start);
-        if (valid) {
-            _prev_kind = mutation_fragment_v2::kind::partition_start;
-            _prev_pos = pos;
-        }
-        return valid;
-    }
-    auto cmp = position_in_partition::tri_compare(_schema);
-    auto res = cmp(_prev_pos, pos);
-    bool valid = true;
-    if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
-        valid = res <= 0;
-    } else {
-        valid = res < 0;
-    }
-    if (valid) {
-        _prev_kind = kind;
-        _prev_pos = pos;
-    }
-    return valid;
-}
-bool mutation_fragment_stream_validator::operator()(mutation_fragment::kind kind, position_in_partition_view pos) {
-    return (*this)(to_mutation_fragment_kind_v2(kind), pos);
-}
 
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment_v2& mf) {
-    const auto valid = (*this)(mf.mutation_fragment_kind(), mf.position());
-    if (valid && mf.is_range_tombstone_change()) {
-        _current_tombstone = mf.as_range_tombstone_change().tombstone();
-    }
-    return valid;
-}
-bool mutation_fragment_stream_validator::operator()(const mutation_fragment& mf) {
-    return (*this)(to_mutation_fragment_kind_v2(mf.mutation_fragment_kind()), mf.position());
-}
+    auto valid = true;
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind) {
-    bool valid = true;
+    // Check fragment kind order
     switch (_prev_kind) {
         case mutation_fragment_v2::kind::partition_start:
-            valid = kind != mutation_fragment_v2::kind::partition_start;
+            valid &= kind != mutation_fragment_v2::kind::partition_start;
```

Comment:
why the change to `&=`?

Reply:
We no longer return as soon as find something invalid so we have to be careful not to overwrite a previous false with a true. Maybe not all these are strictly needed, I just wanted to be careful.

Reply:
Actually, on second look, we don't need the `&=`. We do have a return check after each validation step.

Reply:
> Actually, on second look, we don't need the `&=`. We do have a return check after each validation step.

exactly.